File: neural_network_from_scratch.py
from copy import deepcopy

import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn.datasets import make_moons


def sigmoid(x):
	y = deepcopy(x)
	return 1 / (1 + np.exp(-y))


def sigmoid_prime(x):
	y = sigmoid(x)
	return y * (1 - y)


def tanh(x):
	y = deepcopy(x)
	return np.tan(y)


def tanh_prime(y):
	return 1 - tanh(y) ** 2


def relu(x):
	y = deepcopy(x)
	i = y < 0
	y[i] = 0
	return y


def relu_prime(x):
	y = deepcopy(x)
	i = y < 0
	y[i] = 0
	y[~i] = 1
	return y


def leaky_relu(x):
	y = deepcopy(x)
	i = y < 0.01 * y
	y[i] = 0.01 * y[i]
	return y


def leaky_relu_prime(x):
	y = deepcopy(x)
	i = y < 0
	y[i] = 0.01
	y[~i] = 1
	return y


class NN:
	def __init__(
			self,
			x_train,
			y_train,
			x_test,
			y_test,
	):
		self.X_train = x_train
		self.Y_train = y_train
		self.X_test = x_test
		self.Y_test = y_test

		self.m = None
		self.n0 = None
		self.n1 = None
		self.n2 = None
		self.W1 = None
		self.b1 = None
		self.W2 = None
		self.b2 = None
		self.Z1 = None
		self.A1 = None
		self.Z2 = None
		self.A2 = None
		self.dZ2 = None
		self.dW2 = None
		self.db2 = None
		self.dZ1 = None
		self.dW1 = None
		self.db1 = None
		self.g = None
		self.g_prime = None
		self.cost = None
		self.delta_cost = None
		self.Y_hat_train = None
		self.Y_hat_test = None
		self.alpha = None
		self.max_iterations = None
		# self.min_improvement = None
		self.activation_function = None

	def set_g(self):
		if self.activation_function == 'sigmoid':
			self.g = sigmoid
		elif self.activation_function == 'tanh':
			self.g = np.tanh
		elif self.activation_function == 'relu':
			self.g = relu
		elif self.activation_function == 'leaky_relu':
			self.g = leaky_relu

	def set_g_prime(self):
		if self.activation_function == 'sigmoid':
			self.g_prime = sigmoid_prime
		elif self.activation_function == 'tanh':
			self.g_prime = tanh_prime
		elif self.activation_function == 'relu':
			self.g_prime = relu_prime
		elif self.activation_function == 'leaky_relu':
			self.g_prime = leaky_relu_prime

	def initialize_params(self):
		self.W1 = np.random.randn(self.n1, self.n0) * 0.01
		# self.b1 = np.random.randn(self.n1, 1)
		self.b1 = np.zero(self.n1, 1)
		self.W2 = np.random.randn(self.n2, self.n1) * 0.01
		# self.b2 = np.random.randn(self.n2, 1)
		self.b2 = np.zero(self.n2, 1)

	def forward_propagation(self, X):
		Z1 = np.dot(self.W1, X) + self.b1
		A1 = self.g(Z1)
		Z2 = np.dot(self.W2, A1) + self.b2
		A2 = sigmoid(Z2)
		assert Z1.shape == (self.n1, X.shape[1])
		assert A1.shape == (self.n1, X.shape[1])
		assert Z2.shape == (self.n2, X.shape[1])
		assert A2.shape == (self.n2, X.shape[1])
		return Z1, A1, Z2, A2

	def back_propagation(self):
		self.dZ2 = self.A2 - self.Y_train
		self.dW2 = np.dot(self.dZ2, self.A1.T) / self.m
		self.db2 = np.sum(self.dZ2, axis=1, keepdims=True) / self.m
		self.dZ1 = np.dot(self.W2.T, self.dZ2) * self.g_prime(self.Z1)
		self.dW1 = np.dot(self.dZ1, self.X_train.T)
		self.db1 = np.sum(self.dZ1, axis=1, keepdims=True) / self.m
		assert self.dZ1.shape == self.Z1.shape
		assert self.dW1.shape == self.W1.shape
		assert self.db1.shape == self.b1.shape
		assert self.dZ2.shape == self.Z2.shape
		assert self.dW2.shape == self.W2.shape
		assert self.db2.shape == self.b2.shape

	def calculate_cost(self):
		loss = self.Y_train * np.log(self.A2) + (1 - self.Y_train) * np.log(1 - self.A2)
		sum_loss = np.sum(loss)
		return -sum_loss / self.m

	def predict_prob(self, X):
		Z1, A1, Z2, A2 = self.forward_propagation(X)
		return A2

	def predict_class(self, X):
		probs = self.predict_prob(X)
		return 1 * (probs > 0.5)

	@staticmethod
	def accuracy(Y, Y_hat):
		return np.mean(np.abs(Y_hat - Y))

	def build_model(self,
					n_hidden_nodes=3,
					activation_function='sigmoid',
					alpha=0.01,
					max_iterations=10000,
					# min_improvement=0.00001,
					print_metrics=False
					):

		self.m = self.X_train.shape[1]
		self.n0 = self.X_train.shape[0]
		self.n1 = n_hidden_nodes
		self.n2 = 1

		assert self.X_train.shape == (self.n0, self.m)
		assert self.Y_train.shape == (1, self.m)
		assert self.X_test.shape[0] == self.n0
		assert self.Y_test.shape[0] == 1

		self.activation_function = activation_function
		assert self.activation_function in ['sigmoid', 'tanh', 'relu', 'leaky_relu']

		self.alpha = alpha
		self.max_iterations = max_iterations
		# self.min_improvement = min_improvement

		self.set_g()
		self.set_g_prime()
		self.initialize_params()

		self.cost = []
		self.delta_cost = []

		print('-' * 80)
		print(f'activation function: {self.activation_function}')
		print(f'n hidden nodes: {self.n1}')
		print(f'learning rate: {self.alpha}')
		print(f'n iterations: {max_iterations}')

		# gradient descent -----------------------------------------------------------

		for i in range(self.max_iterations):
			self.Z1, self.A1, self.Z2, self.A2 = self.forward_propagation(self.X_train)
			self.back_propagation()
			self.W1 -= self.alpha * self.dW1
			self.b1 -= self.alpha * self.db1
			self.W2 -= self.alpha * self.dW2
			self.b2 -= self.alpha * self.db2
			self.cost.append(self.calculate_cost())
			self.delta_cost.append(self.cost[i] - self.cost[i-1] if (i > 1) else None)

			if print_metrics:
				if (i+1) % round(self.max_iterations / 10) == 0:
					print(f'Cost after iteration {i+1}: {round(self.cost[i], 5)} (delta: {round(self.delta_cost[i], 10)})')
				if (i+1) == self.max_iterations:
					print(f'stopping at iteration {i+1}: max iterations ({self.max_iterations}) reached')
				# if i > 1000 and self.delta_cost[i] > -self.min_improvement:
				# 	print(f'stopping at iteration {i}: delta cost ({self.delta_cost[i]}) > -min_improvement (-{self.min_improvement})')
				# 	break

		# predictions ----------------------------------------------------------------
		self.Y_hat_train = self.predict_class(self.X_train)
		self.Y_hat_test = self.predict_class(self.X_test)
		self.accuracy(self.Y_train, self.Y_hat_train)
		print(f'cost: {self.cost[-1]}')
		print(f"train accuracy: {100 - self.accuracy(self.Y_train, self.Y_hat_train) * 100} %")
		print(f"test accuracy: {100 - self.accuracy(self.Y_test, self.Y_hat_test) * 100} %")


m_train = 1000
m_test = 200

np.random.seed(42)
X_train_orig, Y_train_orig = make_moons(m_train, noise=0.20)
X_test_orig, Y_test_orig = make_moons(m_test, noise=0.20)

# pd.DataFrame(X_train_orig, columns=['x1', 'x2']).to_csv('data/moons/X_train.csv', index=False)
# pd.DataFrame(X_test_orig, columns=['x1', 'x2']).to_csv('data/moons/X_test.csv', index=False)
# pd.DataFrame(Y_train_orig, columns=['y']).to_csv('data/moons/Y_train.csv', index=False)
# pd.DataFrame(Y_test_orig, columns=['y']).to_csv('data/moons/Y_test.csv', index=False)

Y_train = Y_train_orig.reshape((1, m_train))
Y_test  = Y_test_orig.reshape((1, m_test))
X_train = X_train_orig.T
X_test  = X_test_orig.T

nn = NN(x_train=X_train, y_train=Y_train, x_test=X_test, y_test=Y_test)

nn.build_model(n_hidden_nodes=10, alpha=0.1, max_iterations=10000)  # *
nn.build_model(activation_function='relu', n_hidden_nodes=10, alpha=0.1, max_iterations=10000)  # *
nn.build_model(activation_function='leaky_relu', n_hidden_nodes=10, alpha=0.1, max_iterations=10000)  # *

File: neural_network_from_scratch.R
suppressPackageStartupMessages({
  library(magrittr)
  library(tidyverse)
  library(conflicted)
})
conflicts_prefer(dplyr::filter())
options(dplyr.width = Inf)


# load and reshape data --------------------------------------------------------
train_df = bind_cols(
  read_csv('data/moons/y_train.csv'),
  read_csv('data/moons/X_train.csv'),
)

test_df = bind_cols(
  read_csv('data/moons/y_test.csv'),
  read_csv('data/moons/x_test.csv'),
)

x_train = train_df %>% select(-y)
x_test  = test_df %>% select(-y)
y_train = train_df %>% select(y)
y_test  = test_df %>% select(y)


# plot data + standard lr model ------------------------------------------------
with(train_df, plot(x1, x2, col=y + 1, pch=19, asp=1, main='LR'))

lr = glm(y ~ x1 + x2, data=train_df, family = binomial)
lr_y_hat_train = as.integer(predict(lr, x_train, type='response') > 0.5)
lr_y_hat_test = as.integer(predict(lr, x_test, type='response') > 0.5)
print(str_glue("LR train accuracy: {100 - (mean(abs(lr_y_hat_train - y_train[['y']])) * 100)} %"))
print(str_glue("LR test accuracy: {100 - (mean(abs(lr_y_hat_test - y_test[['y']])) * 100)} %"))

pred_data = with(train_df, tibble(
  x1 = seq(min(x1), max(x1), length.out=100),
  x2 = seq(min(x2), max(x2), length.out=100)
)) %>%
  tidyr::expand(x1, x2)
lr_probs = predict(lr, pred_data, type='response')
lr_preds = as.integer(lr_probs > 0.5)
with(pred_data, points(x=x1, y=x2, col=lr_preds + 1, cex=0.3, pch=15))


# activation functions ---------------------------------------------------------
sigmoid = function(x) {
  1 / (1 + exp(-x))
}

sigmoid_prime = function(x) {
  sigmoid(x) * (1 - sigmoid(x))
}

tanh_prime = function(x) {
  1 - tanh(x)^2
}

relu = function(x) {
  y = x
  i = y < 0
  y[i] = 0
  y
}

relu_prime = function(x) {
  y = x
  i = y < 0
  y[i] = 0
  y[!i] = 1
  y
}

leaky_relu = function(x) {
  # max(0.01 * x, x)
  # if (x < 0.01 * x) 0.01 * x else x
  y = x
  i = y < 0.01 * y
  y[i] = 0.01 * y[i]
  y
}

leaky_relu_prime = function(x) {
  # if (x < 0) 0.01 else 1
  y = x
  i = y < 0
  y[i] = 0.01
  y[!i] = 1
  y
}

g = function(activation_function) {
  activation_function %>%
    switch(
      'sigmoid' = sigmoid,
      'tanh' = tanh,
      'relu' = relu,
      'leaky_relu' = leaky_relu
    )
}

g_prime = function(activation_function) {
  activation_function %>%
    switch(
      'sigmoid' = sigmoid_prime,
      'tanh' = tanh_prime,
      'relu' = relu_prime,
      'leaky_relu' = leaky_relu_prime
    )
}

forward_propagation = function(X, W1, b1, W2, b2, activation_function) {
  Z1 = cbind(b1, W1) %*% rbind(1, X)
  A1 = g(activation_function)(Z1)
  Z2 = cbind(b2, W2) %*% rbind(1, A1)
  A2 = sigmoid(Z2)
  # if (n2 > 1) {
  #   A2 = apply(A2, 2, function(x) x / sum(x))
  # }
  return(list(
    'Z1' = Z1,
    'A1' = A1,
    'Z2' = Z2,
    'A2' = A2
  ))
}

back_propagation = function(X, Y, Z1, A1, W2, Z2, A2, activation_function) {
  m = nrow(X)
  dZ2 = A2 - Y
  dW2 = dZ2 %*% t(A1) / m
  db2 = sum(dZ2) / m
  dZ1 = t(W2) %*% dZ2 * g_prime(activation_function)(Z1)
  dW1 = dZ1 %*% t(X) / m
  db1 = sum(dZ1) / m
  return(list(
    'db1' = db1,
    'dW1' = dW1,
    'db2' = db2,
    'dW2' = dW2
  ))
}

make_predictions = function(model, X) {
  probs = forward_propagation(
    X = X,
    W1 = model[['W1']],
    b1 = model[['b1']],
    W2 = model[['W2']],
    b2 = model[['b2']],
    activation_function = model[['activation_function']]
  )[['A2']]
  as.integer(probs > 0.5)
}

build_model = function(x_train, y_train, x_test, y_test, n_hidden_nodes,
                       activation_function = 'sigmoid',
                       alpha= 0.05, max_iterations=100, min_improvement=0.00001,
                       print_metrics=FALSE) {

  if (0) {
    n_hidden_nodes = 5
    activation_function = 'sigmoid'
    alpha = 0.05
    max_iterations=10
    min_improvement=0.01
    print_metrics=TRUE
  }

  m = nrow(x_train)
  n0 = ncol(x_train)
  n1 = n_hidden_nodes
  n2 = ncol(y_train)

  # reshape data ---------------------------------------------------------------
  Y_train = t(as.matrix(y_train, nrow = 1))
  X_train = t(as.matrix(x_train))
  Y_test = t(as.matrix(y_test, nrow = 1))
  X_test = t(as.matrix(x_test))

  # initialize parameters ------------------------------------------------------
  W1 = matrix(rnorm(n1 * n0), nrow=n1, ncol=n0)
  b1 = matrix(0, nrow=n1, ncol=1)
  W2 = matrix(rnorm(n2 * n1), nrow=n2, ncol=n1)
  b2 = matrix(0, nrow=n2, ncol=1)

  stopifnot(dim(Y_train) == c(1, m))
  stopifnot(dim(X_train) == c(n0, m))
  stopifnot(dim(W1) == c(n1, n0))
  stopifnot(dim(b1) == c(n1, 1))
  stopifnot(dim(W2) == c(n2, n1))
  stopifnot(dim(b2) == c(n2, 1))

  cost = rep(NA, max_iterations)
  delta_cost = rep(NA, max_iterations)

  # gradient descent -----------------------------------------------------------
  for (i in 1:max_iterations) { # i=1

    # forward propagation
    fprop = forward_propagation(X_train, W1, b1, W2, b2, activation_function)
    Z1 = fprop[['Z1']]
    A1 = fprop[['A1']]
    Z2 = fprop[['Z2']]
    A2 = fprop[['A2']]

    stopifnot(dim(Z1) == c(n1, m))
    stopifnot(dim(A1) == c(n1, m))
    stopifnot(dim(Z2) == c(n2, m))
    stopifnot(dim(A2) == dim(Z2))

    cost[i] = -sum(Y_train * log(A2) + (1 - Y_train) * log(1 - A2)) / m

    # back propagation
    bprop = back_propagation(X_train, Y_train, Z1, A1, W2, Z2, A2, activation_function)
    dW1 = bprop[['dW1']]
    db1 = bprop[['db1']]
    dW2 = bprop[['dW2']]
    db2 = bprop[['db2']]

    stopifnot(dim(dW1) == dim(W1))
    stopifnot(dim(db1) == dim(b1))
    stopifnot(dim(dW2) == dim(W2))
    stopifnot(dim(db2) == dim(b2))

    # # Add regularization terms (b1 and b2 don't have regularization terms)
    # dW2 += reg_lambda * W2
    # dW1 += reg_lambda * W1

    # gradient descent parameter update
    W1 = W1 - alpha * dW1
    b1 = b1 - alpha * db1
    W2 = W2 - alpha * dW2
    b2 = b2 - alpha * db2

    cost[i] = -sum(Y_train * log(A2) + (1 - Y_train) * log(1 - A2)) / m

    delta_cost[i] = if (i > 1) cost[i] - cost[i-1] else NA

    if (print_metrics & (i==1 | i %% 100 == 0)) {
      print(str_glue("Cost after iteration {i}: {round(cost[i], 5)} (delta: {round(delta_cost[i], 5)})"))
    }

    if (i == max_iterations) {
      print(str_glue('stopping at iteration {i}: max iterations ({max_iterations}) reached'))
      break
    }

    if (i > 1000) {
      if (delta_cost[i] > -min_improvement) {
        print(str_glue('stopping at iteration {i}: delta cost ({delta_cost[i]}) > -min_improvement (-{min_improvement})'))
        break
      }
    }

  }

  model = list(
    'W1' = W1,
    'b1' = b1,
    'W2' = W2,
    'b2' = b2,
    cost = cost[1:i],
    delta_cost = delta_cost[1:i],
    max_iterations=max_iterations,
    activation_function=activation_function,
    alpha=alpha,
    n_hidden_nodes=n_hidden_nodes
  )

  # predictions ----------------------------------------------------------------
  Y_hat_train = make_predictions(model, X_train)
  Y_hat_test  = make_predictions(model, X_test)

  # print metrics --------------------------------------------------------------
  if (print_metrics) {
    print(str_glue("NN train accuracy: {100 - (mean(abs(Y_hat_train - Y_train)) * 100)} %"))
    print(str_glue("NN test accuracy: {100 - (mean(abs(Y_hat_test - Y_test)) * 100)} %"))
  }

  return(model)

}

# debugonce(tanh)
# debugonce(tanh_prime)
# debugonce(build_model)

set.seed(42)
nn = build_model(
  x_train=x_train,
  y_train=y_train,
  x_test=x_test,
  y_test=y_test,
  n_hidden_nodes=10,
  # activation_function='sigmoid',
  # activation_function='tanh',
  activation_function='relu',
  # activation_function='leaky_relu',
  alpha=0.001,
  max_iterations=10000,
  print_metrics=TRUE
)

with(nn, plot(seq_along(cost), cost, type='l', main='cost'))

nn_preds = make_predictions(nn, t(as.matrix(pred_data)))
title = with(nn, str_glue('NN: {max_iterations} iters, {activation_function}, alpha {alpha}, {n_hidden_nodes} hidden nodes'))
with(train_df, plot(x1, x2, col=y + 1, pch=19, asp=1, main=title))
with(pred_data, points(x=x1, y=x2, col=(as.numeric(nn_preds) > 0.5) + 1, cex=0.3, pch=15))

print(head(cbind(cost=nn[['cost']], delta=nn[['delta_cost']])))
print(tail(cbind(cost=nn[['cost']], delta=nn[['delta_cost']])))


